### PR TITLE
Dockerfile for training on AI Platform, example cluster config files and helper functions

### PR DESCRIPTION
**Pull request by werkaaa**
Dockerfile, cluster config files and helper functions.

- **Dockerfile** with image specification for training on AI Platform.

- Example **cluster config files**.

- Example code (_trainer/ddsp_run_multiple_vms.py_) and configuration gin files from Magenta DDSP repository. In this file I changed the strategy call (**line: 177**) and added copying last operative_config from storage bucket (**line: 135**). 

- Library internal file _trainer/train_util.py_. Early stopping was added (**lines: 198-202**). 

- **_get_strategy()_ function** in the _helper_functions.py_ file. This function is needed for calling MultiWorkerMirroredStrategy.

- **_copy_config_file_from_gstorage()_** function in the _helper_functions.py_ file. this function is needed for copying the configuration file for restoring the training.

Please ignore _utils/run_jobs.py_ script for a while.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -4,8 +4,9 @@ WORKDIR /root
 # Installs sndfile library for reading and writing audio files 
 RUN apt-get update && apt-get install -y libsndfile-dev
 
-# Instals ddsp library and upgrade tensorflow
-RUN pip install --upgrade pip && pip install --upgrade tensorflow && pip install --upgrade ddsp
+# Instals Magenta DDSP library and upgrade Tensorflow
```
Comment (marmolejo) [inline Dockerfile]: In scripts or function headers we would normally use the present tense. Like "Installs" and "upgrades".

File: Dockerfile [hunk 2]
```diff
@@ -41,18 +46,15 @@ RUN mkdir /root/trainer/gin/datasets
...
 
-RUN mkdir /root/tmp
-RUN mkdir /root/tmp/ddsp
+# RUN mkdir /root/tmp
+# RUN mkdir /root/tmp/ddsp
 
+# These parameters can be also specified as part of the job submission arguments
 ENTRYPOINT ["python", "trainer/ddsp_run_multiple_vms.py", "--mode=train", \
      "--alsologtostderr", \
      "--save_dir=gs://werror-2020.appspot.com/mvp/dist", \
```
Comment (marmolejo) [inline Dockerfile]: Here and below. We should parametrize this. People shouldn't be seeing your personal bucket names in the variable contents.

File: Dockerfile [hunk 3]
```diff
@@ -41,18 +46,15 @@ RUN mkdir /root/trainer/gin/datasets
 COPY ./trainer/gin/datasets/tfrecord.gin /root/trainer/gin/datasets/tfrecord.gin
 COPY ./trainer/gin/datasets/base.gin /root/trainer/gin/datasets/base.gin
 
-RUN mkdir /root/tmp
-RUN mkdir /root/tmp/ddsp
+# RUN mkdir /root/tmp
+# RUN mkdir /root/tmp/ddsp
```
Comment (marmolejo) [inline Dockerfile]: With version control it doesn't make sense to keep code comments.

File: trainer/ddsp_run_multiple_vms.py [hunk 4]
```diff
@@ -59,12 +59,15 @@
 from absl import logging
 from ddsp.training import eval_util
 from ddsp.training import models
-from ddsp.training import train_util
+# from ddsp.training import train_util
```
Comment (marmolejo) [inline trainer/ddsp_run_multiple_vms.py]: Here and below: same comments as before: with source control we never keep code comments.

File: trainer/helper_functions.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import json
```
Comment (marmolejo) [inline trainer/helper_functions.py:4]: If you are using headers, you should write them on all files.

File: trainer/train_util.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
+# Copyright 2020 The DDSP Authors.
```
Comment (marmolejo) [inline trainer/train_util.py]: I guess this file is not from the Google style pylint, as I can see 2 spaces and 2 blank lines between functions. In this situation when you want to take files from different sources and don't want to change them to use our formatting rules, you should keep them in a separate directory with a clear readme pointing to the origin.

**Review by marmolejo — APPROVED** [review]

File: Dockerfile (hunk 1 above)
Comment (werkaaa) [reply Dockerfile]: Thank you for pointing this out. I've changed it.

File: Dockerfile (hunk 3 above)
Comment (werkaaa) [reply Dockerfile]: Done

File: trainer/ddsp_run_multiple_vms.py (hunk 4 above)
Comment (werkaaa) [reply trainer/ddsp_run_multiple_vms.py]: Done

File: Dockerfile (hunk 2 above)
Comment (werkaaa) [reply Dockerfile]: Done

File: trainer/train_util.py (hunk 6 above)
Comment (werkaaa) [reply trainer/train_util.py]: Okay, I will restructure it a little and move all internal Magenta files to the separate folder inside the trainer directory.

File: trainer/helper_functions.py (hunk 5 above)
Comment (werkaaa) [reply trainer/helper_functions.py:4]: Sure, I will add header here as well.

**Pull request merged**